### ISSUE TITLE
How many nodes are needed to great a neural network?

### ISSUE THREAD

**Issue by mithi** [mithi/ai-playground#1]
# 1. How many nodes are needed to great a neural network? 

Machine Learning Researchers have used trial and error and experimentation to see that works and what doesn't. Then other machine researchers built on top of that to improve existing neural network architectures. Technically, 

I highly recommend that you check out [this article (neural network zoo)](http://www.asimovinstitute.org/neural-network-zoo/) to familiarize yourself with the most common neural networks used to day and the corresponding problems that they''re solving. 

There is also a paper [(Hrushikesh Mhaskar, Tomaso Poggio)Deep vs. shallow networks : An approximation theory perspective](https://arxiv.org/abs/1608.03287) which argues why Deep (IE many layers between the input and output layers) are better than having shallow layers only. 

Deep neural networks are popular because, in theory, they can learn at different levels of abstraction.
For example, layers near the input could learn simple features like lines and curves of different orientations. The middle/hidden layers can use these as inputs to distinguish more complex shapes like face parts (eyes, nose, etc) and the layers near the output could use that information to recognize specific faces based on facial structures. It is very hard to do transfer learning if you only have a small number of hidden layers (IE shallow networks).

You can also use existing pretrained layers of your deep network for other other purposes (layers that detect different kinds of edges for example) This is called transfer learning. You can learn more about transfer learning [here (machine learning mastery)](https://machinelearningmastery.com/transfer-learning-for-deep-learning/) and [here Dipanjan (DJ) Sarkar](https://towardsdatascience.com/a-comprehensive-hands-on-guide-to-transfer-learning-with-real-world-applications-in-deep-learning-212bf3b2f27a)

**Comment by mithi** [comment]
# 2. What kind of activation function we have to use?(e.g. sigmoid, tanh, ReLU)

We use activation functions to include non-linearity to our neural network. 

It depends on what you are trying to achieve and also trial and experimentation. All of them have pros and cons. Here's a brief summary of some types of [(SAGAR SHARMA)activation functions](https://towardsdatascience.com/activation-functions-neural-networks-1cbd9f8d91d6).

Here is another article that explains which activation functions are "hot right now" [Anish Singh Walia](https://towardsdatascience.com/activation-functions-and-its-types-which-is-better-a9a5310cc8f)

Some other articles that discuss activation functions: 
- [(Dabal Pedamonti)Comparison of non-linear activation functions for deep neural networks onMNIST classification task](https://arxiv.org/pdf/1804.02763.pdf)
- [ML Cheatsheets](https://ml-cheatsheet.readthedocs.io/en/latest/activation_functions.html)
- [CS231n Convolutional Neural Networks for Visual Recognition](http://cs231n.github.io/neural-networks-1/#actfun)
- [(Rinat Maksutov) Activation functions](https://towardsdatascience.com/deep-study-of-a-not-very-deep-neural-network-part-2-activation-functions-fd9bd8d406fc)

One of my favorites activation functions that I use is ELU which has been experimentally proven to produce better results than RELU, but is much more computationally expensive than RELU
- [Exponential Linear Units are the New Cool](http://saikatbasak.in/sigmoid-vs-relu-vs-elu/)
- [(Sefik Ilkin Serengil)ELU](https://sefiks.com/2018/01/02/elu-as-a-neural-networks-activation-function/)

Selu is elu with a little twist. Learn more about it here:
- (SELU: Makes fnn great again)[https://towardsdatascience.com/selu-make-fnns-great-again-snn-8d61526802a9]

**Comment by mithi** [comment]
# 3. What is the exact process of backpropagation?

### A great video by 3Blue1Brown 
- [What is backpropation really doing?](https://www.youtube.com/watch?v=Ilg3gGewQ5U)
- You should also watch the next video "Backpropagation calculus"

### Article
- [Matt Mazur:  A step by step backpropagation](https://mattmazur.com/2015/03/17/a-step-by-step-backpropagation-example/)

**Comment by mithi** [comment]
# Linear and Logistic Regression 
- Logistic Regressions are usually used from classification problems. It's better to read about how different sources describe the difference, the links listed below is a start. 

- [StackOverflow](https://stackoverflow.com/questions/12146914/what-is-the-difference-between-linear-regression-and-logistic-regression)
[Understanding Logistic Regression](https://towardsdatascience.com/understanding-logistic-regression-9b02c2aec102)
- [Difference between Linear Regression and Logistic Regression](https://www.listendata.com/2014/11/difference-between-linear-regression.html)
- [Linear vs. Logistic Probability Models: Which is Better, and When?](https://statisticalhorizons.com/linear-vs-logistic)
- [Difference Between Linear and Logistic Regression](https://techdifferences.com/difference-between-linear-and-logistic-regression.html)

**Comment by mithi** [comment]
# Supervised vs Unsupervised learning 
- supervised is when you know and tell answer to a question that your machine learning model is trying to understand (IE Is this a hotdog or is it not a hotdog?). Supervised learning works as if your machine learning model is a child and you're teaching it to know the difference between a picture of a hotdog and picture that isn't. So what you do is you give the pictures of hotdog (and pictures of non hot dogs) where each picture is labeled. Then the model tries to understand what patterns make something a hotdog. 

- nonsupervised on the other hand is that you give it a bunch of data with no labels and the model tries to find underlying patterns in the data. For example you work at amazon and there are many different shapes and dimensions of each item that you ship. Since the items are varying in sizes, it's impractical to make a box with unique dimensions tailored to each item, so you make 3 types of boxes, small, medium, and large. Given the dimension of the items being shipped, what are the best dimensions for type of box so as not to be wasteful? (we don't want most of our shipments to be inside a box that is mostly filled with air or worst, the item is too large to fit in the largest box we have. 
So we can use machine learning to find patterns to be able to choose which dimensions of boxes are best to produce. 

- [article from geeks for geeks](https://www.geeksforgeeks.org/supervised-unsupervised-learning/)

**Comment by mithi** [comment]
How many nodes we need to create a neural network?
    What kind of activation function we have to use?(e.g. sigmoid, tanh, ReLU)
    What is the exact process of backpropagation?
    What is regressions(linear, logical) used for? And how regressions work?
    What is supervised and unsupervised learning used for?
    What is difference between classification, regression or recommended systems?
    When we use a convolution neural network with keras, we use neural network?
    What are the basics of natural language processing?
